### PR TITLE
add persisted watcher library

### PR DESCRIPTION
**Pull request by wdbaruni**
# Watcher Library

## Overview

The Watcher Library is an internal component of the Bacalhau project that provides a robust event watching and processing system. It's designed to efficiently store, retrieve, and process events. The library ensures events are stored in a durable, ordered manner, allowing for consistent and reliable event processing. It supports features like checkpointing, filtering, and long-polling, while maintaining the ability to replay events from any point in the event history.

## Key Features

1. **Ordered Event Processing**: Events are processed in the exact order they were created, ensuring consistency and predictability in event handling.
2. **Durability**: Events are stored persistently in BoltDB, ensuring they survive system restarts or crashes.
3. **Replayability**: The system allows replaying events from any point in history, facilitating data recovery, debugging, and system reconciliation.
4. **Concurrency**: Multiple watchers can process events concurrently, improving system throughput.
5. **Filtering**: Watchers can filter events based on object types and operations, allowing for targeted event processing.
6. **Checkpointing**: Watchers can save their progress and resume from where they left off, enhancing reliability and efficiency.
7. **Long-polling**: Efficient event retrieval with support for long-polling, reducing unnecessary network traffic and database queries.
8. **Garbage Collection**: Automatic cleanup of old events to manage storage while maintaining the ability to replay from critical points.
9. **Flexible Event Iteration**: Different types of iterators for various use cases, including the ability to start from the oldest event, the latest event, or any specific point in the event history.

## Key Components

1. **Registry**: Manages multiple watchers and provides methods to create and manage watchers.
2. **Watcher**: Represents a single event watcher that processes events sequentially.
3. **EventStore**: Responsible for storing and retrieving events, with BoltDB as the default implementation.
4. **EventHandler**: Interface for handling individual events.
5. **Serializer**: Handles the serialization and deserialization of events.

### Event

An `Event` represents a single occurrence in the system. It has the following properties:

- `SeqNum`: A unique, sequential identifier for the event.
- `Operation`: The type of operation (Create, Update, Delete).
- `ObjectType`: The type of object the event relates to.
- `Object`: The actual data associated with the event.
- `Timestamp`: When the event occurred.

### EventStore

The `EventStore` is responsible for persisting events and providing methods to retrieve them. It uses BoltDB as the underlying storage engine and supports features like caching, checkpointing, and garbage collection.

### Registry

The `Registry` manages multiple watchers. It's the main entry point for components that want to subscribe to events.

### Watcher

A `Watcher` represents a single subscriber to events. It processes events sequentially and can be configured with filters and checkpoints.

### EventIterator

An `EventIterator` defines the starting position for reading events. There are four types of iterators:

1. **TrimHorizonIterator**: Starts from the oldest available event.
2. **LatestIterator**: Starts from the latest available event.
3. **AtSequenceNumberIterator**: Starts at a specific sequence number.
4. **AfterSequenceNumberIterator**: Starts after a specific sequence number.

## Usage

Here's how you typically use the Watcher library within Bacalhau:

1. Create an EventStore:

```go
db, _ := bbolt.Open("events.db", 0600, nil)
store, _ := boltdb.NewEventStore(db)
```

2. Create a Registry:
```go
registry := watcher.NewRegistry(store)
```

3. Implement an EventHandler:
```go
type MyHandler struct{}

func (h *MyHandler) HandleEvent(ctx context.Context, event watcher.Event) error {
    // Process the event
    return nil
}
```

4. Start watching for events:
```go
watcher, _ := registry.Watch(ctx, "my-watcher", &MyHandler{}, 
    watcher.WithFilter(watcher.EventFilter{
        ObjectTypes: []string{"Job", "Execution"},
        Operations: []watcher.Operation{watcher.OperationCreate, watcher.OperationUpdate},
    }),
)
```

5. Store events:
```go
store.StoreEvent(ctx, watcher.OperationCreate, "Job", jobData)
```

### Watch Configuration

When creating a watcher, you can configure it with various options:

- `WithInitialEventIterator(iterator EventIterator)`: Sets the starting position for watching if no checkpoint is found.
- `WithFilter(filter EventFilter)`: Sets the event filter for watching.
- `WithBufferSize(size int)`: Sets the size of the event buffer.
- `WithBatchSize(size int)`: Sets the number of events to fetch in each batch.
- `WithInitialBackoff(backoff time.Duration)`: Sets the initial backoff duration for retries.
- `WithMaxBackoff(backoff time.Duration)`: Sets the maximum backoff duration for retries.
- `WithMaxRetries(maxRetries int)`: Sets the maximum number of retries for event handling.
- `WithRetryStrategy(strategy RetryStrategy)`: Sets the retry strategy for event handling.

Example:

```go
watcher, err := registry.Watch(ctx, "my-watcher", &MyHandler{},
    watcher.WithInitialEventIterator(watcher.TrimHorizonIterator()),
    watcher.WithFilter(watcher.EventFilter{
        ObjectTypes: []string{"Job", "Execution"},
        Operations: []watcher.Operation{watcher.OperationCreate, watcher.OperationUpdate},
    }),
    watcher.WithBufferSize(1000),
    watcher.WithBatchSize(100),
    watcher.WithMaxRetries(3),
    watcher.WithRetryStrategy(watcher.RetryStrategyBlock),
)
```

### EventStore Configuration (BoltDB)

The BoltDB EventStore can be configured with various options:

- `WithEventsBucket(name string)`: Sets the name of the bucket used to store events.
- `WithCheckpointBucket(name string)`: Sets the name of the bucket used to store checkpoints.
- `WithEventSerializer(serializer watcher.Serializer)`: Sets the serializer used for events.
- `WithCacheSize(size int)`: Sets the size of the LRU cache used to store events.
- `WithLongPollingTimeout(timeout time.Duration)`: Sets the timeout duration for long-polling requests.
- `WithGCAgeThreshold(threshold time.Duration)`: Sets the age threshold for event pruning.
- `WithGCCadence(cadence time.Duration)`: Sets the interval at which garbage collection runs.
- `WithGCMaxRecordsPerRun(max int)`: Sets the maximum number of records to process in a single GC run.
- `WithGCMaxDuration(duration time.Duration)`: Sets the maximum duration for a single GC run.

Example:

```go
store, err := boltdb.NewEventStore(db,
    boltdb.WithEventsBucket("myEvents"),
    boltdb.WithCheckpointBucket("myCheckpoints"),
    boltdb.WithCacheSize(1000),
    boltdb.WithLongPollingTimeout(10*time.Second),
)
```

## Best Practices

1. Use meaningful watcher IDs to easily identify different components subscribing to events.
2. Implement error handling in your `EventHandler` to ensure robust event processing.
3. Use appropriate filters to minimize unnecessary event processing.
4. Regularly checkpoint your watchers to enable efficient restarts.
5. Monitor watcher stats to ensure they're keeping up with event volume.

## Troubleshooting

1. If a watcher is falling behind, consider increasing the batch size or optimizing the event handling logic.
2. For performance issues, check the BoltDB file size and consider tuning the garbage collection parameters.

## Future Improvements
1. Enhanced monitoring and metrics.

### CODE REVIEW

**Review by frrist — COMMENTED** [review]
good stuff.

left a couple drive-by comments.

File: pkg/lib/watcher/watcher.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+func (w *watcher) Start() {
+	w.mu.Lock()
+	if w.state != StateIdle && w.state != StateStopped {
+		log.Warn().Str("watcher_id", w.id).Str("state", string(w.state)).
+			Msg("watcher already running/stopped, skipping start")
+		w.mu.Unlock()
+		return
+	}
+
+	w.ctx, w.cancel = context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply pkg/lib/watcher/watcher.go]: Thanks for pointing this out. I understand the concern about storing the provided context and the bad practice mentioned in the link. Though keep in mind I am not storing a _provided_ context as I am creating a fresh one inside Start. My goal is to have better control over how and when background tasks are stopped, which context cancellation doesn't provide.

In this case, the `Start()` method doesn't accept a context. Instead, I create a new context inside `Start` to manage the lifecycle of internal calls and to exit any pending `GetEvents` or `HandleEvent` calls. Initially, I used a stop channel and a fresh background context in each loop, with the stop channel canceling the context. But this was more complicated than necessary.

I realized I only needed to store the cancel method, not the context itself. Fixed in https://github.com/bacalhau-project/bacalhau/pull/4307/commits/e6fd1c94d6d9129a8ff6261d3652e9897120bdeb

File: pkg/lib/watcher/boltdb/boltdb.go [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+		stopGC:   make(chan struct{}),
+	}
+
+	err = errors.Join(
+		validate.NotNil(db, "boltDB instance cannot be nil"),
+		options.validate(),
+	)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply pkg/lib/watcher/boltdb/boltdb.go]: Correct. Fixed in https://github.com/bacalhau-project/bacalhau/pull/4307/commits/f74da4deb3939e189174e8736a20cf11554d6828

File: pkg/lib/watcher/boltdb/boltdb.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-s.notifyCh:
+			// New events might be available, loop and check again
+			// Drain any additional notifications
+			for len(s.notifyCh) > 0 {
+				<-s.notifyCh
+			}
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply pkg/lib/watcher/boltdb/boltdb.go:211]: Added a comment in https://github.com/bacalhau-project/bacalhau/pull/4307/commits/f74da4deb3939e189174e8736a20cf11554d6828

> notifyCh is a channel for notifying watchers of new events.
> GetEvents will block on this channel when no events are immediately available,
> or will return empty events after a long-polling timeout.

File: pkg/lib/watcher/test/eventstore_wrapper.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
+package test
```
Comment (udsamani) [inline pkg/lib/watcher/test/eventstore_wrapper.go:1]: More of a clarification ! Why is this "test" package ?

**Review by udsamani — APPROVED** [review]
Looks Good To Me :) Great Job documenting it ! Really clear for a long PR.

**Pull request merged**